### ISSUE TITLE
Refactoring: Split compiler sensor into GCC and MSVC compiler to use it at the same time

### ISSUE THREAD

**Issue by em2er** [SonarOpenCommunity/sonar-cxx#1180]
**sonar.cxx.compiler.parser** option allows either GCC or Visual C++, not both.

**Comment by guwirth** [comment]
@em2er up to now users were using the one or the other tool chain. Someone else need this feature?

@em2er what is the reason for you to use both in parallel?

**Comment by em2er** [comment]
We have to build cross-platform code (Linux and Windows), so we are using GCC to achieve this. But under Windows we also want: 
- to take an advantage of using powerful MSVC static analyzer (cl.exe /analyze), 
- to prove that our code is compiler-independent by using MSVC (and clang++ actually)
I have written a small script to convert from MSVC output to GCC to make things work.  Here is conversion results:
From MSVC output:
**C:\...\....cpp(83): warning C4189: 'y': local variable is initialized but not referenced**
To GCC output:
**C:\...\....cpp:83:1: warning: MSVC(C4189) 'y': local variable is initialized but not referenced [-Weffc++]**
As you can see, I have to spoof scanner by adding  [-Weffc++] at the end. which is misleading.

**Comment by guwirth** [comment]
See no reason why we should not split it into a GCC and a VC sensor. Any other opinions?
@em2er in case we decide to do it are you able to do it?

**Comment by em2er** [comment]
@guwirth I think it will be nice to keep sonar.cxx.compiler option for backward compatibility marked as deprecated. And add splitted sonar.cxx.msvc.reportPath, sonar.cxx.gcc.reportPath
 > in case we decide to do it are you able to do it?

Hmm, never coded in Java before. I would like to try, but I can start after 18 Jul 2017, is it possible?

**Comment by jeremfg** [comment]
Question:

Could it be made more generic? We'd love to add support for IAR as well via a custom regex.
Currently we catch GCC warnings through sonar-cxx, IAR warnigns through a Jenkins plugin.

We'd love to catch everything using sonar so builds break locally (on dev machine), not just on the build machine.

**Comment by guwirth** [comment]
@jeremfg I think #1081 is what you are meaning with more generic?
* for XML reports we have the generic `sonar.cxx.other.reportPath`: https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Supported-configuration-properties#sonarcxxotherreportpath
* please also read: https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Extending-the-code-analysis

What you can do/try already today is to set a customer regex and rules:
```
sonar.cxx.compiler.parser=...
sonar.cxx.compiler.regex=...
sonar.cxx.compiler-gcc.customRules=... (or sonar.cxx.compiler-vc.customRules)
````

**Comment by jeremfg** [comment]
@guwirth Souds about right with #1081.
It's the fact that GCC and MSVC are built-in that I missed and confused me, thus why this issue exists. IAR would need an extension by one of the two methods you mention.
Thanks for clarifying!

**Comment by guwirth** [comment]
@jeremfg 
> It's the fact that GCC and MSVC are built-in that I missed and confused me

There are only predefined rules. With `customRules` you can define your own rules.

**Comment by guwirth** [comment]
Hi,

Like to split the compiler sensor into two: vc and gcc. 

Remove:
```
sonar.cxx.compiler.parser
sonar.cxx.compiler.reportPath
sonar.cxx.compiler.regex
sonar.cxx.compiler.charset
```
Add:
```
sonar.cxx.vc.reportPath
sonar.cxx.vc.regex
sonar.cxx.vc.charset

sonar.cxx.gcc.reportPath
sonar.cxx.gcc.regex
sonar.cxx.gcc.charset
```

Does someone need the old sensor?
How should we handle the metric? Keep one compiler metric or also split it?

Regards

**Comment by ivangalkin** [comment]
Hi @guwirth,

#1521 de-facto splits both sensors, however they can not be enabled at the same time yet (see `CxxCompilerGccSensor.java` and `CxxCompilerVcSensor.java`). The metric calculation/persisting is done in `CxxIssuesReportSensor`. This implementation doesn't allow sharing of the same metric key between sensors: the next sensor would override the measurements of the previous one.

Although it's possible to accumulate metrics, I would prefer to split the metrics too. It's much clearer to have the 1:1 relationship e.g. 1 sensor = 1 rule repository (#1521), 1 sensor = 1 metric etc.
